commerce/domain/tax/cart/calculator/classes/TaxCartCalculatorSample.cls

```diff
@@ -1,349 +1,242 @@
-      // any invalid scenario found then return.
-      boolean isCvoPresent = false;
-      for (String cartItemId : cartItemById.keySet()) {
-        TaxDataFromExternalService taxDetails = dataFromExternalService.get(cartItemId);
-        if (taxDetails == null) {
-          // add cvo
-          CartExtension.CartValidationOutput cvo = new CartExtension.CartValidationOutput(
-            CartExtension.CartValidationOutputTypeEnum.TAXES,
-            CartExtension.CartValidationOutputLevelEnum.INFO
-          );
-          cvo.setMessage('No tax rates configured for this location.');
-          cartValidationOutputCollection.add(cvo);
-          isCvoPresent = true;
-        }
-      }
-      if (isCvoPresent == true)
-        return;
-
-      for (String cartItemId : dataFromExternalService.keySet()) {
-        TaxDataFromExternalService taxDetailsToCartId = dataFromExternalService.get(cartItemId);
-        CartExtension.CartItem cartItem = cartItemById.get(cartItemId);
-
-        // NOTE: DELETED items get filtered out in the DtoCollection and if there is no tax setup
-        // against any cart item, then that's considered an invalid scenario and added to CVO. If
-        // cart tax numbers are changed that indicates the cart item was MODIFIED, then:
-        // 1. Delete existing and create new cart tax entries in cart item and cart item
-        // adjustments.
-        // 2. Update cart item tax information. Currently, we do not support taxes on tier
-        // adjustment in an extension.
-        boolean isCartItemModified = false;
-        if (
-          (cartItem.getNetUnitPrice() != null &&
-          cartItem.getNetUnitPrice() != taxDetailsToCartId.getNetUnitPrice()) ||
-          !VerifyAdjustmentUpdate(cartItem, taxDetailsToCartId)
-        ) {
-          if (cartItem.getCartTaxes().size() > 0) {
-            cartItem.getCartTaxes().remove(cartItem.getCartTaxes().get(0));
-          }
-          Iterator<CartExtension.CartItemPriceAdjustment> cartItemPriceAdjustmentsIterator = cartItem.getCartItemPriceAdjustments().iterator();
-          while(cartItemPriceAdjustmentsIterator.hasNext()) {
-            CartExtension.CartTaxList cipaTaxes = cartItemPriceAdjustmentsIterator.next()
-              .getCartTaxes();
-            if (cipaTaxes.size() > 0) {
-              cipaTaxes.remove(cipaTaxes.get(0));
+                if (cartItem.getType() == CartExtension.SalesItemTypeEnum.PRODUCT) {
+                	String cartItemId = (cartItem.getId() == null) ? String.valueOf(++cartItemIdSeq) : cartItem.getId();
+                	cartItemByIdMap.put(cartItemId, cartItem);
+                } else if (cartItem.getType() == CartExtension.SalesItemTypeEnum.CHARGE) {
+                	// Shipping cart items are uniquely identified using delivery group id.
+                	CartExtension.CartDeliveryGroup deliveryGroup = cartItem.getCartDeliveryGroup();
+                	chargeItemByDeliveryGroupIdMap.put(deliveryGroup.getId(), cartItem);
```

Comment:
while constructing this map not every cart item will be kept in this map only charge type cart items are kept in this map and also since this will contain only one value in case if there is null also it should be fine right?

Reply:
In case DeliveryGroup has more than one CartItem of type Charge - you aren't going to miss Tax Calculations for one of them?

Reply:
there won't be more than one cartItem of type Charge for one delivery group
cc: @ParameshOddepally

Reply:
I'm pretty sure other Calculators executing before Taxes can easily create more than one `Charge`